File: plugins/mqtt/src/main/java/eu/geekhome/mqttplugin/MoquetteBroker.java
package eu.geekhome.mqttplugin;

import eu.geekhome.services.mqtt.MqttBrokerService;
import eu.geekhome.services.mqtt.MqttListener;
import io.moquette.BrokerConstants;
import io.moquette.broker.ClientDescriptor;
import io.moquette.broker.Server;
import io.moquette.broker.config.MemoryConfig;
import io.moquette.interception.AbstractInterceptHandler;
import io.moquette.interception.InterceptHandler;
import io.moquette.interception.messages.InterceptConnectMessage;
import io.moquette.interception.messages.InterceptConnectionLostMessage;
import io.moquette.interception.messages.InterceptPublishMessage;
import io.netty.buffer.Unpooled;
import io.netty.handler.codec.mqtt.MqttMessageBuilders;
import io.netty.handler.codec.mqtt.MqttPublishMessage;
import io.netty.handler.codec.mqtt.MqttQoS;
import org.pf4j.Extension;

import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Properties;

@Extension
public class MoquetteBroker implements MqttBrokerService {

    private final Server _mqttBroker;
    private final List<MqttListener> _listeners = new ArrayList<>();

    @Override
    public void addMqttListener(MqttListener listener) {
        _listeners.add(listener);
    }

    @Override
    public void removeMqttListener(MqttListener listener) {
        _listeners.remove(listener);
    }

    public MoquetteBroker() {
        _mqttBroker = new Server();
    }

    static class PublisherListener extends AbstractInterceptHandler {

        private final List<MqttListener> _listeners;
        private Server _server;

        PublisherListener(List<MqttListener> listeners, Server server) {
            _listeners = listeners;
            _server = server;
        }

        @Override
        public String getID() {
            return "EmbeddedLauncherPublishListener";
        }

        @Override
        public void onPublish(InterceptPublishMessage msg) {

            int size = 0;
            byte[] buffer = new byte[255];
            while (msg.getPayload().isReadable()) {
                buffer[size] = msg.getPayload().readByte();
                size++;
            }
            String msgAsString = new String(buffer, 0, size);

            System.out.println("Received on topic: " + msg.getTopicName() + ", content: " + msgAsString);

            for (MqttListener listener : _listeners) {
                listener.onPublish(msg.getClientID(), msg.getTopicName(), msgAsString);
            }
        }

        @Override
        public void onConnectionLost(InterceptConnectionLostMessage msg) {
            String clientID = msg.getClientID();
            for (MqttListener listener : _listeners) {
                listener.onDisconnected(clientID);
            }
        }

        @Override
        public void onConnect(InterceptConnectMessage msg) {
            super.onConnect(msg);
            for (ClientDescriptor client : _server.listConnectedClients()) {
                if (msg.getClientID().equals(client.getClientID())) {
                    for (MqttListener listener : _listeners) {
                        try {
                            InetAddress address = InetAddress.getByName(client.getAddress());
                            listener.onConnected(address);
                        } catch (UnknownHostException ignored) {
                        }
                    }
                }
            }
            _server.listConnectedClients();
        }
    }

    public void start() throws IOException {
        List<? extends InterceptHandler> userHandlers = Collections.singletonList(new PublisherListener(_listeners, _mqttBroker));

        MemoryConfig memoryConfig = new MemoryConfig(new Properties());
        memoryConfig.setProperty(BrokerConstants.ALLOW_ANONYMOUS_PROPERTY_NAME, "true");

        _mqttBroker.startServer(memoryConfig, userHandlers);
    }

    public void stop() {
        _mqttBroker.stopServer();
    }

    @Override
    public void publish(String topic, String content) {
        MqttPublishMessage message = MqttMessageBuilders.publish()
                .topicName(topic)
                .retained(true)
                .qos(MqttQoS.EXACTLY_ONCE)
                .payload(Unpooled.copiedBuffer(content.getBytes()))
                .build();

        _mqttBroker.internalPublish(message, "geekHOME Server 2");
    }
}

File: plugins/mqtt/src/main/java/eu/geekhome/mqttplugin/MoquetteBroker.kt
package eu.geekhome.mqttplugin

import eu.geekhome.domain.mqtt.MqttBrokerService
import eu.geekhome.domain.mqtt.MqttListener
import io.moquette.interception.AbstractInterceptHandler
import io.moquette.interception.messages.InterceptPublishMessage
import io.moquette.interception.messages.InterceptConnectionLostMessage
import io.moquette.interception.messages.InterceptConnectMessage
import java.net.InetAddress
import kotlin.Throws
import java.io.IOException
import io.moquette.interception.InterceptHandler
import io.moquette.broker.config.MemoryConfig
import io.moquette.BrokerConstants
import io.moquette.broker.Server
import io.netty.handler.codec.mqtt.MqttMessageBuilders
import io.netty.handler.codec.mqtt.MqttQoS
import io.netty.buffer.Unpooled
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.pf4j.Extension
import java.net.UnknownHostException
import java.util.*

@Extension
class MoquetteBroker : MqttBrokerService {
    private val broker: Server = Server()
    private val listeners: MutableList<MqttListener> = ArrayList()
    var pluginScope: CoroutineScope? = null


    override fun addMqttListener(listener: MqttListener) {
        listeners.add(listener)
    }

    override fun removeMqttListener(listener: MqttListener) {
        listeners.remove(listener)
    }

    internal class PublisherListener(
        private val pluginScope: CoroutineScope,
        private val listeners: List<MqttListener>,
        private val server: Server
    ) :
        AbstractInterceptHandler() {
        override fun getID(): String {
            return "EmbeddedLauncherPublishListener"
        }

        override fun onPublish(msg: InterceptPublishMessage) {
            var size = 0
            val buffer = ByteArray(255)
            while (msg.payload.isReadable) {
                buffer[size] = msg.payload.readByte()
                size++
            }
            val msgAsString = String(buffer, 0, size)
            println("Received on topic: " + msg.topicName + ", content: " + msgAsString)
            for (listener in listeners) {
                listener.onPublish(msg.clientID, msg.topicName, msgAsString)
            }
        }

        override fun onConnectionLost(msg: InterceptConnectionLostMessage) {
            val clientID = msg.clientID
            for (listener in listeners) {
                listener.onDisconnected(clientID)
            }
        }

        @Suppress("BlockingMethodInNonBlockingContext")
        override fun onConnect(msg: InterceptConnectMessage) {
            super.onConnect(msg)
            pluginScope.launch {
                for (client in server.listConnectedClients()) {
                    if (msg.clientID == client.clientID) {
                        for (listener in listeners) {
                            try {
                                val address = InetAddress.getByName(client.address)
                                listener.onConnected(address)
                            } catch (ignored: UnknownHostException) {
                            }
                        }
                    }
                }
            }
        }
    }

    @Throws(IOException::class)
    fun start() {
        if (pluginScope != null) {
            pluginScope?.cancel("MQTT Broker already started")
        }
        pluginScope = CoroutineScope(Dispatchers.IO)
        val userHandlers: List<InterceptHandler?> = listOf(PublisherListener(pluginScope!!, listeners, broker))
        val memoryConfig = MemoryConfig(Properties())
        memoryConfig.setProperty(BrokerConstants.ALLOW_ANONYMOUS_PROPERTY_NAME, "true")
        broker.startServer(memoryConfig, userHandlers)
    }

    fun stop() {
        pluginScope?.cancel("MQTT Broker has been stopped")
        pluginScope = null
        broker.stopServer()
    }

    override fun publish(topic: String?, content: String) {
        val message = MqttMessageBuilders.publish()
            .topicName(topic)
            .retained(true)
            .qos(MqttQoS.EXACTLY_ONCE)
            .payload(Unpooled.copiedBuffer(content.toByteArray()))
            .build()
        broker.internalPublish(message, "geekHOME Server 2")
    }

}